lib/puppet/face/facts.rb

```diff
@@ -87,4 +87,44 @@
       nil
     end
   end
+
+  action(:show) do
+    summary _("Facter plugin sync")
+    arguments _("[<facts>]")
+    description <<-'EOT'
+    Reads facts from the local system facter gem.
+    EOT
+    returns "The output of facter with added puppet specific facts"
+    notes <<-'EOT'
+
+    EOT
+    examples <<-'EOT'
+    retrieve facts:
+
+    $ puppet facts show os
+    EOT
+
+    render_as :json
+
+    when_invoked do |*args|
+      facts_to_add = {clientcert: Puppet.settings[:certname], clientversion: Puppet.version.to_s, clientnoop: Puppet.settings[:noop]}
+      Puppet.settings.preferred_run_mode = :agent
+      Facter.reset
+
+      Puppet::Node::Facts.indirection.terminus_class = :facter
+      certname = Puppet.settings[:certname]
+      Puppet::Node::Facts.indirection.search_paths(certname, options)
```

Comment:
Can we call `Puppet::Node::Facts.indirection.find(certname)`, since the [facter terminus already calls `add_local_facts`](https://github.com/puppetlabs/puppet/blob/e2a883df7e8bd35238cd274e71dd2d99d6af62ee/lib/puppet/indirector/facts/facter.rb#L36), and that [adds local facts](https://github.com/puppetlabs/puppet/blob/e2a883df7e8bd35238cd274e71dd2d99d6af62ee/lib/puppet/node/facts.rb#L29-L33)?

Reply:
`add_local_facts` method doesn't add the facts to facter (with `Facter.add` method ) instead it adds the facts to the `Puppet::Node::Facts.values` hash. On `show` action an object of type `Puppet::Node::Facts` is no longer created, in order to add the local facts, i added them directly to Facter (using `Facter.add` method)

Reply:
Hey @oanatmaria, sorry, I think I'm missing something. Could we just call `Puppet::Node::Facts.indirection.find(Puppet[:certname])` and render the returned `Puppet::Node::Facts#values`?

I think a current limitation (both before and after your changes) is that you can't write a custom or external fact that depends on the `clientcert`, `clientversion`, `clientnoop` facts, since they're added to the facts object after resolution. But I think that's ok, since it's been that way since those facts were added.